ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/KsqlMaterialization.java

```diff
@@ -128,148 +131,157 @@ public MaterializedWindowedTable windowed() {
+    }
+  }
+
   final class KsqlMaterializedTable implements MaterializedTable {
 
-    private final MaterializedTable table;
+    private final StreamsMaterializedTable table;
 
-    KsqlMaterializedTable(final MaterializedTable table) {
+    KsqlMaterializedTable(final StreamsMaterializedTable table) {
       this.table = requireNonNull(table, "table'");
     }
 
     @Override
-    public KsMaterializedQueryResult<Row> get(final GenericKey key, final int partition) {
-      if (transforms.isEmpty()) {
-        return table.get(key, partition);
-      }
+    public KsqlMaterializedQueryResult<Row> get(
+        final GenericKey key,
+        final int partition,
+        final Optional<ConsistencyOffsetVector> consistencyVector
+    ) {
+      final Optional<Position> position = consistencyVector
+          .map(offsetVector -> Position.fromMap(offsetVector.getOffsetVector()));
+      final KsMaterializedQueryResult<Row> result = table.get(key, partition, position);
+      updateConsistencyVector(consistencyVector, result.getPosition());
 
-      final KsMaterializedQueryResult<Row> result = table.get(key, partition);
-      final Iterator<Row> iterator = Streams.stream(result.getRowIterator())
-          .map(row -> filterAndTransform(row.key(), getIntermediateRow(row), row.rowTime())
-              .map(v -> row.withValue(v, schema())))
-          .filter(Optional::isPresent)
-          .map(Optional::get)
-          .iterator();
-
-      if (result.getPosition().isPresent()) {
-        return KsMaterializedQueryResult.rowIteratorWithPosition(
-            iterator, result.getPosition().get());
+      if (transforms.isEmpty()) {
+        return KsqlMaterializedQueryResult.rowIterator(result.getRowIterator());
       } else {
-        return KsMaterializedQueryResult.rowIterator(iterator);
+        final Iterator<Row> iterator = Streams.stream(result.getRowIterator())
+            .map(row -> filterAndTransform(row.key(), getIntermediateRow(row), row.rowTime())
+                .map(v -> row.withValue(v, schema())))
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .iterator();
+        return KsqlMaterializedQueryResult.rowIterator(iterator);
       }
     }
 
     @Override
-    public KsMaterializedQueryResult<Row> get(final int partition) {
-      if (transforms.isEmpty()) {
-        return table.get(partition);
-      }
-
-      final KsMaterializedQueryResult<Row> result = table.get(partition);
-
-      final Iterator<Row> iterator = Streams.stream(result.getRowIterator())
-          .map(row -> filterAndTransform(row.key(), getIntermediateRow(row), row.rowTime())
-              .map(v -> row.withValue(v, schema())))
-          .filter(Optional::isPresent)
-          .map(Optional::get)
-          .iterator();
+    public KsqlMaterializedQueryResult<Row> get(
+        final int partition,
+        final Optional<ConsistencyOffsetVector> consistencyVector
```

Comment:
I'm wondering (this can be done in a separate PR) if instead of using optional everywhere if it was possible to use a no-op consistency vector. tends to make the code a little easier to read and encapsulation a little clearer but definitely not a big deal

then we could make methods like `updateConsistencyVector` part of the no-op class' implementation of `updateFromPosition` (it just does nothing)

Reply:
Yes, this is definitely something we were thinking about. Patrick mentioned it as well. It will make the code cleaner for sure